### PR TITLE
Switching filters specification in API to allow for arbitrary

### PR DESCRIPTION
**Pull request by danqrodney**
and, ors, not recursively.

Adding NOT_EQUAL operator.

### CODE REVIEW

**Comment by danqrodney** [comment]
This isn't remotely done yet -- just wanted to get your feedback on the filter API changes.

File: api/src/main/resources/client_api.yaml [hunk 1]
```diff
@@ -190,6 +184,36 @@ definitions:
...
+    type: object
+    description: >
+      A list of filters applied to the results of a query. Only results matching the filter criteria
+      should be returned.
+    properties:
+      not:
+        description: >
+         Set to true if a result matching allOf or anyOf below should result in a result *not*
+         being returned.
+      allOf:
```
Comment (calbach) [inline api/src/main/resources/client_api.yaml]: I'm assuming exactly one of these three fields should be populated?

File: api/src/main/resources/client_api.yaml [hunk 2]
```diff
@@ -190,6 +184,36 @@ definitions:
...
+        description: >
+          A list of result filters. All filters matching means a result should be returned (or not returned
+          if "not" is true.)
+        type: array
+        items:
+          $ref: "#/definitions/ResultFilters"
```
Comment (calbach) [inline api/src/main/resources/client_api.yaml]: Hopefully the various Swagger codegens can handle this properly, worth checking things look OK for Python at least.

File: api/src/main/resources/client_api.yaml [hunk 3]
```diff
@@ -247,4 +271,4 @@ definitions:
 
   Operator:
     type: string
-    enum: [EQUAL, LESS_THAN, GREATER_THAN, LESS_THAN_OR_EQUAL_TO, GREATER_THAN_OR_EQUAL_TO, LIKE, IN]
+    enum: [EQUAL, NOT_EQUAL, LESS_THAN, GREATER_THAN, LESS_THAN_OR_EQUAL_TO, GREATER_THAN_OR_EQUAL_TO, LIKE, IN]
```
Comment (freemabd1) [inline api/src/main/resources/client_api.yaml]: So what happens when i combine IN operator with allOf filter?

File: api/src/bigquerytest/java/org/pmiops/workbench/cohorts/CohortMaterializationServiceTest.java [hunk 4]
```diff
@@ -550,6 +573,45 @@ public void testMaterializeCohortPersonFieldSetPersonIdWithStringIsNull() {
...
+    TableQuery tableQuery = new TableQuery();
+    tableQuery.setTableName("person");
+    tableQuery.setColumns(ImmutableList.of("person_id"));
+    ColumnFilter filter = new ColumnFilter();
+    filter.setColumnName("ethnicity_source_value");
+    filter.setOperator(Operator.NOT_EQUAL);
+    filter.setValueNull(true);
+    tableQuery.addFiltersItem(ImmutableList.of(filter));
+    FieldSet fieldSet = new FieldSet();
+    fieldSet.setTableQuery(tableQuery);
```
Comment (freemabd1) [inline api/src/bigquerytest/java/org/pmiops/workbench/cohorts/CohortMaterializationServiceTest.java]: So this doesn't have to do with the filters but why do we need the FieldSet? Currently it's just a wrapper for TableQuery that doesn't add any extra value. Could we just pass TableQuery instead or am i missing some added functionality that will appear later?

File: api/src/main/resources/client_api.yaml (hunk 1 above)
Comment (danqrodney) [reply api/src/main/resources/client_api.yaml]: Yeah, added a comment.

File: api/src/main/resources/client_api.yaml (hunk 2 above)
Comment (danqrodney) [reply api/src/main/resources/client_api.yaml]: Will check this.

File: api/src/main/resources/client_api.yaml (hunk 3 above)
Comment (danqrodney) [reply api/src/main/resources/client_api.yaml]: The operator is a field on ColumnFilter, not ResultFilters... you would make a ResultFilter containing a ColumnFilter with IN if you want an in clause on some column.

"IN" can be used on ColumnFilter with values or valueNumbers to make an in clause.

File: api/src/bigquerytest/java/org/pmiops/workbench/cohorts/CohortMaterializationServiceTest.java (hunk 4 above)
Comment (danqrodney) [reply api/src/bigquerytest/java/org/pmiops/workbench/cohorts/CohortMaterializationServiceTest.java]: In future I'm expecting we're going to have other types of queries under FieldSet (e.g. for fetching annotations, or particular views of the data that don't correspond to OMOP tables.) So having FieldSet at the top level allows us to add other children under it.

File: api/src/main/resources/client_api.yaml (hunk 2 above)
Comment (danqrodney) [reply api/src/main/resources/client_api.yaml]: Generated Python looks fine to me.

**Comment by calbach** [comment]
Is this ready for review?

**Comment by danqrodney** [comment]
Yes please!

File: api/src/main/resources/client_api.yaml [hunk 5]
```diff
@@ -190,6 +184,37 @@ definitions:
         items:
           type: string
 
+  ResultFilters:
```
Comment (calbach) [inline api/src/main/resources/client_api.yaml]: opt: perhaps this should just be called `Predicate` or similar at this point as it now support arbitrary boolean expressions

Comment (danqrodney) [reply api/src/main/resources/client_api.yaml]: While I agree "predicate" sorta makes sense, it feels more geeky / less researcher-friendly to me than "filter". 

I also wanted some prefix to make it clear this applies to the results that come back and to distinguish it from other places in our API that might use predicates / filters.

(I'm not strongly attached to the name, just explaining how I landed on this.)

**Pull request merged**